### ISSUE TITLE
P4_16 program struct field name collision causes wrong bmv2 JSON file

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/p4c#486]
See the sample P4_16 program and JSON file generated on this issue: https://github.com/p4lang/behavioral-model/issues/331

I do not know whether fixing this requires only a change in p4c-bm2-ss, or whether it would require changes in both that and bmv2.

**Comment by antoninbas** [comment]
I had pushed a patch to bmv2 to check for name / id duplicates when appropriate (https://github.com/p4lang/behavioral-model/pull/334). However it broke the tests for this repo (see output below). I reverted my change temporarily, but it would be great if the bmv2 backend could be fixed.

```
FAIL: bmv2/testdata/p4_16_samples/action_selector_shared-bmv2.p4
================================================================
Calling target program-options parser
Duplicate actions with name 'drop' in action profile 'as'
Could not connect to thrift client on port 9795
Make sure the switch is running and that you have the right port
Check for  ../testdata/p4_16_samples/action_selector_shared-bmv2.stf
*** simple_switch died with return code 1
*** CLI process failed with exit code 1
*** Test failed
FAIL bmv2/testdata/p4_16_samples/action_selector_shared-bmv2.p4.test (exit status: 1)
FAIL: bmv2/testdata/p4_16_samples/x-bmv2.p4
===========================================
Calling target program-options parser
Duplicate objects of type 'header' with name 's'
Could not connect to thrift client on port 9822
Make sure the switch is running and that you have the right port
Check for  ../testdata/p4_16_samples/x-bmv2.stf
*** simple_switch died with return code 1
*** CLI process failed with exit code 1
*** Test failed
FAIL bmv2/testdata/p4_16_samples/x-bmv2.p4.test (exit status: 1)
FAIL: bmv2/testdata/p4_14_samples/queueing.p4
=============================================
Calling target program-options parser
Duplicate objects of type 'header type' with name 'queueing_metadata_t'
Could not connect to thrift client on port 9850
Make sure the switch is running and that you have the right port
Check for  ../testdata/p4_14_samples/queueing.stf
*** simple_switch died with return code 1
*** CLI process failed with exit code 1
*** Test failed
FAIL bmv2/testdata/p4_14_samples/queueing.p4.test (exit status: 1)
FAIL: bmv2/testdata/p4_14_samples/16-TwoReferences.p4
=====================================================
Calling target program-options parser
Duplicate objects of type 'control node' with name 'C'
Could not connect to thrift client on port 9359
Make sure the switch is running and that you have the right port
Check for  ../testdata/p4_14_samples/16-TwoReferences.stf
*** simple_switch died with return code 1
*** CLI process failed with exit code 1
*** Test failed
FAIL bmv2/testdata/p4_14_samples/16-TwoReferences.p4.test (exit status: 1)
```

**Comment by mbudiu-vmw** [comment]
We can mangle the names in the output, but I wonder whether this will affect the control-plane.

**Comment by mbudiu-vmw** [comment]
Indeed, that's a problem.
For example, if any of these fields are used as table keys, they will no longer match.
Unfortunately the @name annotations for keys do not seem to be represented in the JSON, so we cannot change the key names. @sethfowler @ChrisDodd @antoninbas : do you have any suggestions on how to fix this?

**Comment by sethfowler** [comment]
@antoninbas, can we just replace the names we're using in the JSON right now with fully qualified names? Does BMV2 actually interpret these names or does it just treat them as opaque?

**Comment by antoninbas** [comment]
You can use the fully-qualified names in the JSON, but you would have to use the same names in the STF files.

**Comment by sethfowler** [comment]
That'd be fine with me. Anyone see any downsides?

**Comment by jafingerhut** [comment]
I think I'm hitting this issue also with an action declared at the top level in a P4_16 program, used as the action for two different tables.  That might just be another instance of the same issue, solved the same way, but I thought I would mention it.  I can create a small sample test program if it would be useful.

**Comment by jafingerhut** [comment]
Just a note that this issue may be the same as, or at least is related to, this one: https://github.com/p4lang/p4c/issues/455

**Comment by jafingerhut** [comment]
Confirmed today that this issue also exists if you have a P4_16 program with the same named table in two different control blocks.  p4c produces a JSON file that can cause bmv2 to crash.  I have confirmed that the reverted change https://github.com/p4lang/behavioral-model/pull/334 to behavioral-model code can catch this issue, too.

**Comment by jafingerhut** [comment]
Just wanted to ping on this issue, to see if there was any interest someone working on a fix.

The symptoms when someone writes a P4-16 program that triggers this issue are highly non-obvious and difficult to debug.  You compile your program with p4c-bm2-ss, you start simple_switch or some other bmv2 executable on the JSON file, and it either immediately crashes, or it crashes when certain packets are sent in, maybe only after certain table entries are added.  There is no obvious correlation to anything you wrote in your program at all.  It is entirely mystifying what causes the crash.

**Comment by jafingerhut** [comment]
@mbudiu-vmw @sethfowler @antoninbas Would it be reasonable to re-commit these extra checks back into bmv2, and mark these 4 failing p4c test cases as XFAIL until the p4c bmv2 backend is corrected?

**Comment by jafingerhut** [comment]
I could easily be missing something, but it seems that changing names so that they have extra 'prefixes' that make them unique would be a pretty straightforward way to avoid the duplicates.

e.g. instead of 'ipv4.fieldname' there would be 'meta.ipv4.fieldname' or 'hdr.ipv4.fieldname' of the bmv2 JSON generated by the example program causing problems, attached to issue https://github.com/p4lang/behavioral-model/issues/331

**Comment by jafingerhut** [comment]
Of those 4 programs that you list that gave an error when you tried out these bmv2 changes in Apr 2017:
```
FAIL: bmv2/testdata/p4_16_samples/action_selector_shared-bmv2.p4
Duplicate actions with name 'drop' in action profile 'as'

FAIL: bmv2/testdata/p4_16_samples/x-bmv2.p4
Duplicate objects of type 'header' with name 's'

FAIL: bmv2/testdata/p4_14_samples/queueing.p4
Duplicate objects of type 'header type' with name 'queueing_metadata_t'

FAIL: bmv2/testdata/p4_14_samples/16-TwoReferences.p4
Duplicate objects of type 'control node' with name 'C'
```
queueing.p4 no longer generates a JSON file with duplicate header type 'queueing_metadata_t', probably because of some change to p4c-bm2-ss since April to now.  So that one need not even be marked as XFAIL if these bmv2 checks are re-committed now.

x-bmv2.p4 and 16-TwoReferences.p4 seem to be cases where the p4c-bm2-ss bmv2 backend is generating JSON that violates the assumptions of bmv2, and it would be good to figure out how to change the bmv2 backend not to use those duplicate names any more.

For action_selector_shared-bmv2.p4, it seems to me that the only 'duplicates' here are that the same action selector 'as' is used by two different tables, and the two tables have the same set of action names.  That is exactly the restriction that some people want to impose for two tables that share the same action selector, yes?  For that one, it seems that these extra checks are giving an error when they should not.

**Comment by mbudiu-vmw** [comment]
This needs help from the control-plane, which needs to use the correct unambiguous names.
In particular, this will require us to rewrite some STF files.
I don't have a problem to do that, but other back-ends may need to do the same thing.

**Comment by mbudiu-vmw** [comment]
I agree, let's mark these as XFAIL and insert the stricter checks in bmv2. This is the only way we can solve this problem.

**Comment by jafingerhut** [comment]
I would like to find out from @antoninbas whether his added checks that cause these errors:
```
FAIL: bmv2/testdata/p4_16_samples/x-bmv2.p4
Duplicate objects of type 'header' with name 's'
```
are actual restrictions of the bmv2 JSON file.  If they are, I'd like to know how he expects the JSON file ought to be created to avoid violating the new checks.

**Comment by antoninbas** [comment]
@jafingerhut Unlike the other checks for duplicate entries, this particular one may be incorrect.
It seems that this error would be thrown every time an action profile is shared between tables, even tough this is a perfectly valid scenario. Let me verify that and open a new PR without this specific check.

**Comment by antoninbas** [comment]
@jafingerhut This check was indeed incorrect. I opened a new pull request with my amended commit from 6 months ago: jafingerhut

**Comment by antoninbas** [comment]
I merged the PR, the check is now enabled in the master branch of bmv2.

**Comment by mbudiu-vmw** [comment]
I think I have a fix for this which I will submit soon. It's not particularly pretty, although the code changes are not large. Amazingly enough, I didn't have to change too many stf files. (This worries me a bit, in fact.)
The fix may break other control-planes. I have filed an issue with BMv2 to allow in JSON key names to be specified separately from key expressions.

**Comment by jafingerhut** [comment]
I tried the version of p4c after the commit for #1106 was merged in, on the couple of test programs I submitted for this issue, and things definitely look better.  I suspect Antonin Bas has done more thorough checking than I have before that change was merged in.  Closing this issue.  Will open a different one if anything else arises.   Thanks!
